stdlib/sql.go

```diff
@@ -191,41 +218,102 @@ func GetConnector(config pgx.ConnConfig, opts ...OptionOpenDB) driver.Connector
 	return c
 }
 
+func GetPoolConnector(pool *pgxpool.Pool, opts ...OptionOpenDB) driver.Connector {
+	c := connector{
+		pool:          pool,
+		BeforeAcquire: func(context.Context, *pgxpool.Pool) error { return nil }, // noop before acquire by default
+		AfterAcquire:  func(context.Context, *pgxpool.Conn) error { return nil }, // noop after acquire by default
+		ResetSession:  func(context.Context, *pgx.Conn) error { return nil },     // noop reset session by default
+		driver:        pgxDriver,
+	}
+
+	for _, opt := range opts {
+		opt(&c)
+	}
+
+	return c
+}
+
 func OpenDB(config pgx.ConnConfig, opts ...OptionOpenDB) *sql.DB {
 	c := GetConnector(config, opts...)
 	return sql.OpenDB(c)
 }
 
+func OpenDBFromPool(pool *pgxpool.Pool, opts ...OptionOpenDB) *sql.DB {
+	c := GetPoolConnector(pool, opts...)
+	db := sql.OpenDB(c)
+	db.SetMaxIdleConns(0)
+	return db
+}
+
 type connector struct {
 	pgx.ConnConfig
+	pool          *pgxpool.Pool
 	BeforeConnect func(context.Context, *pgx.ConnConfig) error // function to call before creation of every new connection
 	AfterConnect  func(context.Context, *pgx.Conn) error       // function to call after creation of every new connection
+	BeforeAcquire func(context.Context, *pgxpool.Pool) error   // function to call before acquiring of every new connection
+	AfterAcquire  func(context.Context, *pgxpool.Conn) error   // function to call after acquiring of every new connection
 	ResetSession  func(context.Context, *pgx.Conn) error       // function is called before a connection is reused
 	driver        *Driver
 }
 
 // Connect implement driver.Connector interface
 func (c connector) Connect(ctx context.Context) (driver.Conn, error) {
 	var (
-		err  error
-		conn *pgx.Conn
+		connConfig pgx.ConnConfig
+		conn       *pgx.Conn
+		close      func(context.Context) error
+		err        error
 	)
 
-	// Create a shallow copy of the config, so that BeforeConnect can safely modify it
-	connConfig := c.ConnConfig
-	if err = c.BeforeConnect(ctx, &connConfig); err != nil {
-		return nil, err
-	}
+	if c.pool == nil {
+		// Create a shallow copy of the config, so that BeforeConnect can safely modify it
+		connConfig = c.ConnConfig
 
-	if conn, err = pgx.ConnectConfig(ctx, &connConfig); err != nil {
-		return nil, err
-	}
+		if err = c.BeforeConnect(ctx, &connConfig); err != nil {
+			return nil, err
+		}
 
-	if err = c.AfterConnect(ctx, conn); err != nil {
-		return nil, err
+		if conn, err = pgx.ConnectConfig(ctx, &connConfig); err != nil {
+			return nil, err
+		}
+
+		if err = c.AfterConnect(ctx, conn); err != nil {
+			return nil, err
+		}
+
+		close = conn.Close
+	} else {
+		var pconn *pgxpool.Conn
+
+		if err = c.BeforeAcquire(ctx, c.pool); err != nil {
```

Comment:
whats the use case for `sql.Driver` level BeforeAcquire? Pool already has this callbacks internally.

Reply:
This is a good point, fixed!